pkg/workloads/cortex/lib/client/onnx.py

```diff
@@ -12,79 +12,500 @@
+                                    str(e),
+                                )
+                        model, _ = self._models.get_model(model_name, model_version, tag)
+
+        if not self._multiple_processes and self._caching_enabled:
+            # determine model version
+            try:
+                if tag != "":
+                    model_version = self._get_model_version_from_tree(
+                        model_name, tag, self._models_tree.model_info(model_name)
+                    )
+            except ValueError:
+                # if model_name hasn't been found
+                raise UserRuntimeException(
+                    f"'{model_name}' model of tag {model_version} wasn't found in the list of available models"
+                )
+
+            # grab shared access to model tree
+            available_model = True
+            with LockedModelsTree(self._models_tree, "r", model_name, model_version):
+
+                # check if the versioned model exists
+                model_id = model_name + "-" + model_version
+                if model_id not in self._models_tree:
+                    available_model = False
+                    raise WithBreak
+
+                # retrieve model tree's metadata
+                upstream_model = self._models_tree[model_id]
+                current_upstream_ts = int(upstream_model["timestamp"].timestamp())
+
+            if not available_model:
+                return None
+
+            # grab shared access to models holder and retrieve model
+            update_model = False
+            with LockedModel(self._models, "r", model_name, model_version):
+                status, upstream_ts = self._models.has_model(model_name, model_version)
+                if status in ["not-available", "on-disk"] or (
+                    status != "not-available"
+                    and upstream_ts < current_upstream_ts
+                    and not (status == "in-memory" and model_name in self._spec_local_model_names)
+                ):
+                    update_model = True
+                    raise WithBreak
+                model, _ = self._models.get_model(model_name, model_version, tag)
+
+            # download, load into memory the model and retrieve it
+            if update_model:
+                # grab exclusive access to models holder
+                with LockedModel(self._models, "w", model_name, model_version):
+
+                    # check model status
+                    status, upstream_ts = self._models.has_model(model_name, model_version)
+
+                    # refresh disk model
+                    if model_name not in self._spec_local_model_names and (
+                        status == "not-available"
+                        or (
+                            status in ["on-disk", "in-memory"] and upstream_ts < current_upstream_ts
+                        )
+                    ):
+                        # remove model from disk and memory
```

Comment:
client.python._get_model seemed to have some helpful logging statements that aren't here. I think we should copy them here as well (after the comments have been updated of course)